### PR TITLE
Fixing HazelcastTransactionManager timeout

### PR DESCRIPTION
**Pull request by dbaltor**
HazelcastTransactionManager doesn't use timeout values set either for the transaction or as default value in the transaction manager

Fixes https://github.com/hazelcast/hazelcast/issues/23460

Backport: 
[v5.3.0-DEVEL-5](https://github.com/hazelcast/hazelcast/releases/tag/v5.3.0-DEVEL-5)
[v5.2.1](https://github.com/hazelcast/hazelcast/releases/tag/v5.2.1)
[v5.0.4](https://github.com/hazelcast/hazelcast/releases/tag/v5.0.4)
[v4.2.6](https://github.com/hazelcast/hazelcast/releases/tag/v4.2.6)
[v4.1.10](https://github.com/hazelcast/hazelcast/releases/tag/v4.1.10)

Breaking changes (list specific methods/types/messages):
* None

Checklist:
- [ x ] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [ x ] Label `Add to Release Notes` or `Not Release Notes content` set
- [ x ] Send backports/forwardports if fix needs to be applied to past/future releases

### CODE REVIEW

**Comment by vbekiaris** [comment]
run-lab-run

**Comment by dbaltor** [comment]
>run-lab-run

Addressed checkstyle errors. Could someone run the pipeline again please?

**Comment by vbekiaris** [comment]
run-lab-run

**Comment by dbaltor** [comment]
> run-lab-run
@vbekiaris  could you give it another whirl please?

**Comment by dbaltor** [comment]
May I ask for somebody run this pipeline again please?

**Comment by vbekiaris** [comment]
run-lab-run

File: hazelcast-spring-tests/src/test/java/com/hazelcast/spring/transaction/TestSpringManagedHazelcastTransaction.java [hunk 1]
```diff
@@ -95,6 +103,63 @@ public void noExceptionWhenTransaction() {
...
+
+        // then
+        assertEquals(1L, instance.getMap("dummyObjectMap").size());
+    }
+
+    /**
+     * Tests that transaction times out when its duration exceeds the value configured for the transaction
+     */
+    @Test
+    public void TransactionTimedOutExceptionWhenTimeoutValueIsSetForTransaction() {
```
Comment (vbekiaris) [inline hazelcast-spring-tests/src/test/java/com/hazelcast/spring/transaction/TestSpringManagedHazelcastTransaction.java:123]:
```suggestion
    public void transactionTimedOutExceptionWhenTimeoutValueIsSetForTransaction() {
```

File: hazelcast-spring-tests/src/test/java/com/hazelcast/spring/transaction/TestSpringManagedHazelcastTransaction.java [hunk 2]
```diff
@@ -95,6 +103,63 @@ public void noExceptionWhenTransaction() {
...
+        service.putWithDelay_transactionTimeoutValue(new DummyObject(1L, "magic"), TIMEOUT + 1);
+    }
+
+    /**
+     * Tests that transaction times out when its duration exceeds the default value
+     * configured in the transaction manager AND no value is configured for the transaction
+     */
+    @Test
+    @DirtiesContext(methodMode = AFTER_METHOD)
+    public void TransactionTimedOutExceptionWhenTimeoutValueIsSetInTransactionManager() {
```
Comment (vbekiaris) [inline hazelcast-spring-tests/src/test/java/com/hazelcast/spring/transaction/TestSpringManagedHazelcastTransaction.java:138]:
```suggestion
    public void transactionTimedOutExceptionWhenTimeoutValueIsSetInTransactionManager() {
```

File: hazelcast-spring-tests/src/test/java/com/hazelcast/spring/transaction/TestSpringManagedHazelcastTransaction.java [hunk 3]
```diff
@@ -95,6 +103,63 @@ public void noExceptionWhenTransaction() {
...
+        // when
+        service.putWithDelay(new DummyObject(1L, "magic"), TIMEOUT + 1);
+    }
+
+    /**
+     * Tests that timeout value of the transaction takes precedence over the default value in the transaction manager
+     */
+    @Test
+    @DirtiesContext(methodMode = AFTER_METHOD)
+    public void TransactionTimeoutTakesPrecedenceOverTransactionManagerDefaultTimeout() {
```
Comment (vbekiaris) [inline hazelcast-spring-tests/src/test/java/com/hazelcast/spring/transaction/TestSpringManagedHazelcastTransaction.java:153]:
```suggestion
    public void transactionTimeoutTakesPrecedenceOverTransactionManagerDefaultTimeout() {
```

**Review by vbekiaris — APPROVED** [review]
thanks for the fix @dbaltor , added minor comments to align test method names with existing ones.

File: hazelcast-spring/src/main/java/com/hazelcast/spring/transaction/HazelcastTransactionManager.java [hunk 4]
```diff
@@ -121,6 +121,26 @@ protected void doBegin(Object transaction, TransactionDefinition definition) thr
...
+                transactionContext = hazelcastInstance.newTransactionContext();
+        } else {
+            TransactionOptions options = new TransactionOptions();
+            if (definition.getTimeout() != TransactionDefinition.TIMEOUT_DEFAULT) {
+                options.setTimeout(definition.getTimeout(), TimeUnit.SECONDS);
+            } else {
+                options.setTimeout(this.getDefaultTimeout(), TimeUnit.SECONDS);
+            }
+            transactionContext = hazelcastInstance.newTransactionContext(options);
+        }
```
Comment (ldziedziul) [inline hazelcast-spring/src/main/java/com/hazelcast/spring/transaction/HazelcastTransactionManager.java:137]: Whaty do you think about simplifying nested if-s? 
```suggestion
        TransactionOptions options = TransactionOptions.getDefault();
        if (definition.getTimeout() != TransactionDefinition.TIMEOUT_DEFAULT) {
            options.setTimeout(definition.getTimeout(), TimeUnit.SECONDS);
        } else if (getDefaultTimeout() != TransactionDefinition.TIMEOUT_DEFAULT) {
            options.setTimeout(getDefaultTimeout(), TimeUnit.SECONDS);
        }
        TransactionContext transactionContext = hazelcastInstance.newTransactionContext(options);
```

**Review by ldziedziul — COMMENTED** [review]
Nice job! left a comment.

File: hazelcast-spring/src/main/java/com/hazelcast/spring/transaction/HazelcastTransactionManager.java
Comment (dbaltor) [inline hazelcast-spring/src/main/java/com/hazelcast/spring/transaction/HazelcastTransactionManager.java:1]: The original logic that I tried to preserve would create a new `transactionContext` only if there was none in place. The suggestion seems to always create a new one. Is that the intended outcome?

Comment (dbaltor) [reply hazelcast-spring/src/main/java/com/hazelcast/spring/transaction/HazelcastTransactionManager.java:1]: I'm talking nonsense here! Please never mind.

**Review by ldziedziul — APPROVED** [review]
LGTM, good work!

**Comment by vbekiaris** [comment]
Merged, thanks @dbaltor for the fix 🏅 !
Can you prepare backports for currently maintained branches? (these are 5.2.z, 5.1.z & 5.0.z)

**Pull request merged**